packages/dds/tree/src/feature-libraries/forest-summary/incrementalSummaryBuilder.ts

```diff
@@ -0,0 +1,410 @@
+import {
+	SummaryTreeBuilder,
+	type ReadAndParseBlob,
+} from "@fluidframework/runtime-utils/internal";
+import { setInNestedMap, tryGetFromNestedMap, type NestedMap } from "../../util/index.js";
+import type {
+	ChunkReferenceId,
+	EncodedFieldBatch,
+	IncrementalEncoderDecoder,
+	TreeChunk,
+} from "../chunked-forest/index.js";
+import type { ITreeCursorSynchronous } from "../../core/index.js";
+import { SummaryType } from "@fluidframework/driver-definitions";
+import type { IChannelStorageService } from "@fluidframework/datastore-definitions/internal";
+import type { ISnapshotTree } from "@fluidframework/driver-definitions/internal";
+import { LoggingError } from "@fluidframework/telemetry-utils/internal";
+
+/**
+ * The key for the summary blob that contains the contents of an incremental chunk.
+ */
+const chunkContentsBlobKey = "contents";
+
+/**
+ * State that tells whether a summary is currently being tracked.
+ */
+export const ForestSummaryTrackingState = {
+	/** A summary is currently being tracked. */
+	Tracking: "Tracking",
+	/** A summary is ready to be tracked. */
+	ReadyToTrack: "ReadyToTrack",
+} as const;
+export type ForestSummaryTrackingState =
+	(typeof ForestSummaryTrackingState)[keyof typeof ForestSummaryTrackingState];
+
+/**
+ * The properties of a chunk that is tracked for every summary. If a chunk doesn't change between summaries,
+ * these properties will be used to generate a summary handle for the chunk.
+ */
+interface ChunkSummaryProperties {
+	/**
+	 * The reference ID of the chunk which uniquely identifies it under its parent's summary tree.
+	 * The summary for this chunk will be stored against this reference ID as key in the summary tree.
+	 */
+	referenceId: ChunkReferenceId;
+	/**
+	 * The path for this chunk's summary in the summary tree relative to the forest's summary tree.
+	 * This path is used to generate a summary handle for the chunk if it doesn't change between summaries.
+	 */
+	summaryPath: string;
+}
+
+/**
+ * The properties of a summary being tracked.
+ */
+interface TrackedSummaryProperties {
+	/**
+	 * The sequence number of the summary in progress.
+	 */
+	readonly summarySequenceNumber: number;
+	/**
+	 * The base path for the latest summary that was successful. This is used to generate summary handles.
+	 */
+	readonly latestSummaryBasePath: string;
+	/**
+	 * Whether the summary being tracked is a full tree summary.
+	 * If true, the summary will not contain any summary handles. All chunks must be summarized in full.
+	 */
+	readonly fullTree: boolean;
+	/**
+	 * Represents the path of a chunk in the summary tree relative to the forest's summary tree. Each item in the
+	 * array is a reference ID of a chunk in the summary tree starting from the chunk under forest summary tree.
+	 * When a chunk is summarized, this array will be used to generate the path for the chunk's summary in the
+	 * summary tree.
+	 */
+	readonly chunkSummaryPath: number[];
```

Comment:
I think this is supposed to be readonly?
```suggestion
	readonly chunkSummaryPath: readonly number[];
```

Reply:
This is updated as we recursively encode chunks - a chunk's reference id is added to this before encoding its child so that it has the proper summary path.